File: sdk/python/foodblock/offline.py
"""
Offline queue for creating FoodBlocks without network connectivity.
Blocks are stored locally and synced when connectivity is restored.

See Section 5.5 of the whitepaper.
"""

import json
from .block import create, update


class OfflineQueue:
    """Offline queue that stores blocks locally for later batch sync."""

    def __init__(self):
        self._blocks = []

    def create(self, type, state=None, refs=None):
        """Create a block and add it to the offline queue."""
        block = create(type, state or {}, refs or {})
        self._blocks.append(block)
        return block

    def update(self, previous_hash, type, state=None, refs=None):
        """Create an update block and add it to the offline queue."""
        block = update(previous_hash, type, state or {}, refs or {})
        self._blocks.append(block)
        return block

    @property
    def blocks(self):
        """Get all queued blocks (copy)."""
        return list(self._blocks)

    @property
    def length(self):
        """Number of queued blocks."""
        return len(self._blocks)

    def clear(self):
        """Clear the queue (e.g. after successful sync)."""
        self._blocks = []

    def sorted(self):
        """
        Sort blocks in dependency order for sync.
        Blocks that reference other blocks in the queue are placed after their dependencies.
        """
        hashes = {b['hash'] for b in self._blocks}
        graph = {}

        for block in self._blocks:
            deps = []
            refs = block.get('refs', {})
            for ref in refs.values():
                ref_hashes = ref if isinstance(ref, list) else [ref]
                for h in ref_hashes:
                    if h in hashes:
                        deps.append(h)
            graph[block['hash']] = deps

        # Topological sort
        visited = set()
        result = []

        def visit(h):
            if h in visited:
                return
            visited.add(h)
            for dep in graph.get(h, []):
                visit(dep)
            block = next((b for b in self._blocks if b['hash'] == h), None)
            if block:
                result.append(block)

        for block in self._blocks:
            visit(block['hash'])

        return result

    def sync(self, url, session=None):
        """
        Sync queued blocks to a remote server.

        Args:
            url: The FoodBlock server URL (e.g. 'http://localhost:3111')
            session: Optional requests.Session or compatible HTTP client.
                     Must have a .post(url, json=...) method.

        Returns:
            dict with inserted, skipped, failed arrays

        Raises:
            ImportError if requests is not installed and no session provided
            Exception on HTTP errors
        """
        if session is None:
            try:
                import requests
                session = requests.Session()
            except ImportError:
                raise ImportError(
                    "Install 'requests' for sync support: pip install requests"
                )

        sorted_blocks = self.sorted()
        response = session.post(
            f"{url}/blocks/batch",
            json={"blocks": sorted_blocks}
        )
        response.raise_for_status()
        result = response.json()
        self.clear()
        return result


def offline_queue():
    """Create a new offline queue."""
    return OfflineQueue()

File: sdk/swift/Sources/Offline.swift
import Foundation

/// Offline queue for creating FoodBlocks without network connectivity.
/// Blocks are stored locally and synced when connectivity is restored.
/// See Section 5.5 of the whitepaper.
public class OfflineQueue {

    private var _blocks: [FoodBlock] = []

    public init() {}

    // MARK: - Block Creation

    /// Create a block and add it to the offline queue.
    ///
    /// - Parameters:
    ///   - type: The block type (e.g. "substance.product")
    ///   - state: The block's state dictionary
    ///   - refs: The block's refs dictionary
    /// - Returns: The created FoodBlock
    @discardableResult
    public func create(type: String, state: [String: Any] = [:], refs: [String: Any] = [:]) -> FoodBlock {
        let block = FoodBlock.create(type: type, state: state, refs: refs)
        _blocks.append(block)
        return block
    }

    /// Create an update block and add it to the offline queue.
    ///
    /// - Parameters:
    ///   - previousHash: The hash of the block being updated
    ///   - type: The block type
    ///   - state: The new state (full replacement)
    ///   - refs: Additional refs (updates ref is added automatically)
    /// - Returns: The created FoodBlock
    @discardableResult
    public func update(previousHash: String, type: String, state: [String: Any] = [:], refs: [String: Any] = [:]) -> FoodBlock {
        let block = FoodBlock.update(previousHash: previousHash, type: type, state: state, refs: refs)
        _blocks.append(block)
        return block
    }

    // MARK: - Queue Access

    /// Get all queued blocks (returns a copy).
    public var blocks: [FoodBlock] {
        return _blocks
    }

    /// Number of queued blocks.
    public var count: Int {
        return _blocks.count
    }

    // MARK: - Queue Management

    /// Clear the queue (e.g. after successful sync).
    public func clear() {
        _blocks = []
    }

    // MARK: - Topological Sort

    /// Sort blocks in dependency order for sync.
    ///
    /// Blocks that reference other blocks in the queue are placed after their dependencies.
    /// Uses a depth-first topological sort: if block B references block A (via any ref),
    /// then A will appear before B in the result.
    ///
    /// - Returns: Array of FoodBlocks sorted so dependencies come first
    public func sorted() -> [FoodBlock] {
        // Collect all hashes in the queue for fast lookup
        let queueHashes = Set(_blocks.map { $0.hash })

        // Build dependency graph: block hash -> [dependency hashes within the queue]
        var graph: [String: [String]] = [:]
        for block in _blocks {
            var deps: [String] = []
            for (_, refValue) in block.refs {
                let refHashes = extractHashes(from: refValue.value)
                for h in refHashes {
                    if queueHashes.contains(h) {
                        deps.append(h)
                    }
                }
            }
            graph[block.hash] = deps
        }

        // Topological sort via DFS
        var visited = Set<String>()
        var result: [FoodBlock] = []

        // Index blocks by hash for fast lookup
        let blocksByHash: [String: FoodBlock] = Dictionary(
            _blocks.map { ($0.hash, $0) },
            uniquingKeysWith: { first, _ in first }
        )

        func visit(_ hash: String) {
            if visited.contains(hash) { return }
            visited.insert(hash)

            // Visit all dependencies first
            for dep in (graph[hash] ?? []) {
                visit(dep)
            }

            // Add this block after its dependencies
            if let block = blocksByHash[hash] {
                result.append(block)
            }
        }

        // Visit all blocks (order of insertion as tiebreaker)
        for block in _blocks {
            visit(block.hash)
        }

        return result
    }

    // MARK: - Helpers

    /// Extract block hashes from a ref value.
    /// Handles both single string refs and array refs.
    private func extractHashes(from value: Any) -> [String] {
        if let str = value as? String {
            return [str]
        }
        if let arr = value as? [String] {
            return arr
        }
        if let arr = value as? [Any] {
            return arr.compactMap { $0 as? String }
        }
        return []
    }
}
